## Building a recomender system for books:

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

### Reading books data: 

In [2]:
books = pd.read_csv('Books.csv')
ratings = pd.read_csv('Ratings.csv')
users = pd.read_csv('Users.csv')

C:\Users\varun\AppData\Local\Temp\ipykernel_26728\378180666.py:1: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  books = pd.read_csv('Books.csv')


In [3]:
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [4]:
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [5]:
ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [6]:
# Lets check the shape of the data:
print(f"books shape : {books.shape}")
print(f"ratings shape : {ratings.shape}")
print(f"users shape: {users.shape}")

books shape : (271360, 8)
ratings shape : (1149780, 3)
users shape: (278858, 3)


### EDA:

**Checking the null values in each dataset:**

In [7]:
books.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            1
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

In [8]:
ratings.isnull().sum()

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

In [9]:
users.isnull().sum()

User-ID          0
Location         0
Age         110762
dtype: int64

Books dataset has very few missing values.

Users dataset has huge number of missing values in the age column.

**Checking for duplicate values:**

In [10]:
books.duplicated().sum()

0

In [11]:
ratings.duplicated().sum()

0

In [12]:
users.duplicated().sum()

0

None of the dataset has the duplicate rows.

In [22]:
ratings['Book-Rating'].value_counts()

0     716109
8     103736
10     78610
7      76457
9      67541
5      50974
6      36924
4       8904
3       5996
2       2759
1       1770
Name: Book-Rating, dtype: int64

### Lets build a poularity based recommender system:

* We will count the total number of ratings for each book.

* Then we will calculate the mean of the ratings.

* we will only consider books having atleat 250 ratings to it.

* Finally recommend top 50 books according to above criteria.

In [13]:
# To do this we will use books and ratings datasets:

ratings_with_name = ratings.merge(books, on ='ISBN')

In [14]:
ratings_with_name.head()

User-ID        ISBN  Book-Rating            Book-Title Book-Author  \
0   276725  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
1     2313  034545104X            5  Flesh Tones: A Novel  M. J. Rose   
2     6543  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
3     8680  034545104X            5  Flesh Tones: A Novel  M. J. Rose   
4    10314  034545104X            9  Flesh Tones: A Novel  M. J. Rose   

  Year-Of-Publication         Publisher  \
0                2002  Ballantine Books   
1                2002  Ballantine Books   
2                2002  Ballantine Books   
3                2002  Ballantine Books   
4                2002  Ballantine Books   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/034545104X.0...   
2  http://images.amazon.com/images/P/034545104X.0...   
3  http://images.amazon.com/images/P/034545104X.0...   
4  http://images.amazon.com/images/P/034545104X.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/034545104X.0...   
2  http://images.amazon.com/images/P/034545104X.0...   
3  http://images.amazon.com/images/P/034545104X.0...   
4  http://images.amazon.com/images/P/034545104X.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/034545104X.0...  
1  http://images.amazon.com/images/P/034545104X.0...  
2  http://images.amazon.com/images/P/034545104X.0...  
3  http://images.amazon.com/images/P/034545104X.0...  
4  http://images.amazon.com/images/P/034545104X.0...

In [15]:
num_rating_df = ratings_with_name.groupby('Book-Title').count()['Book-Rating'].reset_index()
num_rating_df.rename(columns = {'Book-Rating':'num_ratings'}, inplace = True)
num_rating_df

Book-Title  num_ratings
0        A Light in the Storm: The Civil War Diary of ...            4
1                                   Always Have Popsicles            1
2                    Apple Magic (The Collector's series)            1
3        Ask Lily (Young Women of Faith: Lily Series, ...            1
4        Beyond IBM: Leadership Marketing and Finance ...            1
...                                                   ...          ...
241066                                      Ã?Â?lpiraten.            2
241067                     Ã?Â?rger mit Produkt X. Roman.            4
241068                                Ã?Â?sterlich leben.            1
241069                              Ã?Â?stlich der Berge.            3
241070                                  Ã?Â?thique en toc            2

[241071 rows x 2 columns]

In [16]:
avg_rating_df = ratings_with_name.groupby('Book-Title').mean()['Book-Rating'].reset_index()
avg_rating_df.rename(columns = {'Book-Rating':'avg_ratings'}, inplace = True)
avg_rating_df

Book-Title  avg_ratings
0        A Light in the Storm: The Civil War Diary of ...     2.250000
1                                   Always Have Popsicles     0.000000
2                    Apple Magic (The Collector's series)     0.000000
3        Ask Lily (Young Women of Faith: Lily Series, ...     8.000000
4        Beyond IBM: Leadership Marketing and Finance ...     0.000000
...                                                   ...          ...
241066                                      Ã?Â?lpiraten.     0.000000
241067                     Ã?Â?rger mit Produkt X. Roman.     5.250000
241068                                Ã?Â?sterlich leben.     7.000000
241069                              Ã?Â?stlich der Berge.     2.666667
241070                                  Ã?Â?thique en toc     4.000000

[241071 rows x 2 columns]

In [17]:
popular_df = num_rating_df.merge(avg_rating_df, on = 'Book-Title')
popular_df

Book-Title  num_ratings  \
0        A Light in the Storm: The Civil War Diary of ...            4   
1                                   Always Have Popsicles            1   
2                    Apple Magic (The Collector's series)            1   
3        Ask Lily (Young Women of Faith: Lily Series, ...            1   
4        Beyond IBM: Leadership Marketing and Finance ...            1   
...                                                   ...          ...   
241066                                      Ã?Â?lpiraten.            2   
241067                     Ã?Â?rger mit Produkt X. Roman.            4   
241068                                Ã?Â?sterlich leben.            1   
241069                              Ã?Â?stlich der Berge.            3   
241070                                  Ã?Â?thique en toc            2   

        avg_ratings  
0          2.250000  
1          0.000000  
2          0.000000  
3          8.000000  
4          0.000000  
...             ...  
241066     0.000000  
241067     5.250000  
241068     7.000000  
241069     2.666667  
241070     4.000000  

[241071 rows x 3 columns]

In [23]:
popular_df = popular_df[popular_df['num_ratings']>=250].sort_values('avg_ratings', ascending=False).head(50)

In [24]:
popular_df

Book-Title  num_ratings  \
80434   Harry Potter and the Prisoner of Azkaban (Book 3)          428   
80422        Harry Potter and the Goblet of Fire (Book 4)          387   
80441      Harry Potter and the Sorcerer's Stone (Book 1)          278   
80426   Harry Potter and the Order of the Phoenix (Boo...          347   
80414    Harry Potter and the Chamber of Secrets (Book 2)          556   
191612  The Hobbit : The Enchanting Prelude to The Lor...          281   
187377  The Fellowship of the Ring (The Lord of the Ri...          368   
80445   Harry Potter and the Sorcerer's Stone (Harry P...          575   
211384     The Two Towers (The Lord of the Rings, Part 2)          260   
219741                              To Kill a Mockingbird          510   
183573                                  The Da Vinci Code          898   
187880                 The Five People You Meet in Heaven          430   
180556                             The Catcher in the Rye          449   
196326                          The Lovely Bones: A Novel         1295   
764                                                  1984          284   
144165                           Prodigal Summer: A Novel          253   
128670                                         Neverwhere          265   
206502                            The Secret Life of Bees          774   
168719  Stupid White Men ...and Other Sorry Excuses fo...          283   
223135  Tuesdays with Morrie: An Old Man, a Young Man,...          493   
204387                The Red Tent (Bestselling Backlist)          723   
191589               The Hitchhiker's Guide to the Galaxy          268   
129379   Nickel and Dimed: On (Not) Getting By in America          335   
93381                                       Into the Wild          252   
63867                                      Fahrenheit 451          409   
74750                           Girl with a Pearl Earring          526   
136145                                          Outlander          283   
233370  Where the Heart Is (Oprah's Book Club (Paperba...          585   
156102                     Seabiscuit: An American Legend          275   
107962                                         Life of Pi          664   
176845                                     The Bean Trees          389   
2281    A Child Called \It\": One Child's Courage to S...          265   
8434                                       ANGELA'S ASHES          279   
76343                                         Good in Bed          490   
64931   Fast Food Nation: The Dark Side of the All-Ame...          321   
12700                                       American Gods          302   
161645                                 Skipping Christmas          322   
105777  Left Behind: A Novel of the Earth's Last Days ...          318   
189551    The Golden Compass (His Dark Materials, Book 1)          336   
181679                                   The Color Purple          314   
160336                               Silence of the Lambs          256   
8752                                          About a Boy          262   
158138                  Seven Up (A Stephanie Plum Novel)          278   
175097  The Alchemist: A Fable About Following Your Dream          266   
80069   Hard Eight : A Stephanie Plum Novel (A Stephan...          269   
170101                       Suzanne's Diary for Nicholas          457   
111073                                  Lord of the Flies          259   
5664                              A Prayer for Owen Meany          413   
212070   The Vampire Lestat (Vampire Chronicles, Book II)          301   
233851       White Oleander : A Novel (Oprah's Book Club)          356   

        avg_ratings  
80434      5.852804  
80422      5.824289  
80441      5.737410  
80426      5.501441  
80414      5.183453  
191612     5.007117  
187377     4.948370  
80445      4.895652  
211384     4.880769  
219741     4.700000  
183573     4.642539  
187880  

These are top 50 books recommendation to everyone, it is common through out every users.

In [30]:
# Lets merge popular_df with books to get the details about books.

popular_df = popular_df.merge(books, on = 'Book-Title').drop_duplicates('Book-Title')[['Book-Title','Book-Author','Image-URL-M','num_ratings','avg_ratings']]

In [31]:
popular_df.head()

Book-Title    Book-Author  \
0   Harry Potter and the Prisoner of Azkaban (Book 3)  J. K. Rowling   
3        Harry Potter and the Goblet of Fire (Book 4)  J. K. Rowling   
5      Harry Potter and the Sorcerer's Stone (Book 1)  J. K. Rowling   
9   Harry Potter and the Order of the Phoenix (Boo...  J. K. Rowling   
13   Harry Potter and the Chamber of Secrets (Book 2)  J. K. Rowling   

                                          Image-URL-M  num_ratings  \
0   http://images.amazon.com/images/P/0439136350.0...          428   
3   http://images.amazon.com/images/P/0439139597.0...          387   
5   http://images.amazon.com/images/P/0590353403.0...          278   
9   http://images.amazon.com/images/P/043935806X.0...          347   
13  http://images.amazon.com/images/P/0439064872.0...          556   

    avg_ratings  
0      5.852804  
3      5.824289  
5      5.737410  
9      5.501441  
13     5.183453

**THESE ARE TOP 50 BOOKS RECOMMENDED BASED OF AVERAGE RATINGS**

## Collaborative Filtering Based Recommender System:

* We wil try to filter out the users who has voted for more than 250 times, since they have experice in reading books.

* Also we will consider the books which has more than 50 ratings to it, inorder to consider only popular books.


In [32]:
x = ratings_with_name.groupby('User-ID').count()['Book-Rating'] > 200

# This will give all the users who has voted more than 200 times
top_users = x[x].index

In [34]:
len(top_users)

811

There are 811 users

In [35]:
# now lets find the books that these 811 users has rated.

filtered_ratings = ratings_with_name[ratings_with_name['User-ID'].isin(top_users)]

In [37]:
filtered_ratings

User-ID        ISBN  Book-Rating  \
2           6543  034545104X            0   
5          23768  034545104X            0   
7          28523  034545104X            0   
15         77940  034545104X            0   
16         81977  034545104X            0   
...          ...         ...          ...   
1030883   275970  1880837927            0   
1030884   275970  188717897X            0   
1030885   275970  1888889047            0   
1030886   275970  1931868123            0   
1030887   275970  3411086211           10   

                                                Book-Title  \
2                                     Flesh Tones: A Novel   
5                                     Flesh Tones: A Novel   
7                                     Flesh Tones: A Novel   
15                                    Flesh Tones: A Novel   
16                                    Flesh Tones: A Novel   
...                                                    ...   
1030883                         The Theology of the Hammer   
1030884  The Ordeal of Integration: Progress and Resent...   
1030885  Pushcart's Complete Rotten Reviews &amp; Rejec...   
1030886  There's a Porcupine in My Outhouse: Misadventu...   
1030887                                         Die Biene.   

                           Book-Author Year-Of-Publication  \
2                           M. J. Rose                2002   
5                           M. J. Rose                2002   
7                           M. J. Rose                2002   
15                          M. J. Rose                2002   
16                          M. J. Rose                2002   
...                                ...                 ...   
1030883                 Millard Fuller                1994   
1030884              Orlando Patterson                1998   
1030885                 Bill Henderson                1998   
1030886                   Mike Tougias                2002   
1030887  Sybil Gr�¤fin Sch�¶nfeldt                1993   

                                    Publisher  \
2                            Ballantine Books   
5                            Ballantine Books   
7                            Ballantine Books   
15                           Ballantine Books   
16                           Ballantine Books   
...                                       ...   
1030883         Smyth &amp; Helwys Publishing   
1030884                Civitas Book Publisher   
1030885                        Pushcart Press   
1030886                    Capital Books (VA)   
1030887  Bibliographisches Institut, Mannheim   

                                               Image-URL-S  \
2        http://images.amazon.com/images/P/034545104X.0...   
5        http://images.amazon.com/images/P/034545104X.0...   
7        http://images.amazon.com/images/P/034545104X.0...   
15       http://images.amazon.com/images/P/034545104X.0...   
16       http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1030883  http://images.amazon.com/images/P/1880837927.0...   
1030884  http://images.amazon.com/images/P/188717897X.0...   
1030885  http://images.amazon.com/images/P/1888889047.0...   
1030886  http://images.amazon.com/images/P/1931868123.0...   
1030887  http://images.amazon.com/images/P/3411086211.0...   

                                               Image-URL-M  \
2        http://images.amazon.com/images/P/034545104X.0...   
5        http://images.amazon.com/images/P/034545104X.0...   
7        http://images.amazon.com/images/P/034545104X.0...   
15       http://images.amazon.com/images/P/034545104X.0...   
16       http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1030883  http://images.amazon.com/images/P/1880837927.0...   
1030884  http://images.amazon.com/images/P/188717897X.0...   
1030885  http://images.amazon.com/images/P/1888889047.0...   
1030886  http://images.a

In [38]:
y = filtered_ratings.groupby('Book-Title').count()['Book-Rating'] >=50

famous_books = y[y].index

In [40]:
len(famous_books)

706

There are 706 books which has more than 50 ratings to it.

In [41]:
final_ratings = filtered_ratings[filtered_ratings['Book-Title'].isin(famous_books)]

In [42]:
final_ratings

User-ID        ISBN  Book-Rating              Book-Title  \
63        278418  0446520802            0            The Notebook   
65          3363  0446520802            0            The Notebook   
66          7158  0446520802           10            The Notebook   
69         11676  0446520802           10            The Notebook   
74         23768  0446520802            6            The Notebook   
...          ...         ...          ...                     ...   
1026724   266865  0531001725           10  The Catcher in the Rye   
1027923   269566  0670809381            0                  Echoes   
1028777   271284  0440910927            0           The Rainmaker   
1029070   271705  B0001PIOX4            0          Fahrenheit 451   
1030868   275970  1586210661            9  Me Talk Pretty One Day   

                   Book-Author Year-Of-Publication                Publisher  \
63             Nicholas Sparks                1996             Warner Books   
65             Nicholas Sparks                1996             Warner Books   
66             Nicholas Sparks                1996             Warner Books   
69             Nicholas Sparks                1996             Warner Books   
74             Nicholas Sparks                1996             Warner Books   
...                        ...                 ...                      ...   
1026724  Jerome David Salinger                1973   Scholastic Library Pub   
1027923           Maeve Binchy                1986              Penguin USA   
1028777           John Grisham                1995                   Island   
1029070           Ray Bradbury                1993     Simon &amp; Schuster   
1030868          David Sedaris                2001  Time Warner Audio Major   

                                               Image-URL-S  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-M  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-L  
63       http://images.amazon.com/images/P/0446520802.0...  
65       http://images.amazon.com/images/P/0446520802.0...  
66       http://images.amazon.com/images/P/0446520802.0...  
69       http://images.amazon.com/images/P/0446520802.0...  
74       http://images.amazon.com/images/P/0446520802.0...  
...                                                    ...  
1026724  http://images.amazon.com/images/P/0531001725.0...  
1027923  http://images.amazon.com/images/P/0670809381.0...  
1028777  http://images.amazon.com/images/P/0440910927.0...  
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...  
1030868  http://images.amazon.com/images/P/1586210661.0...  

[58586 r

My final_df has details about books whihc has been rated more than 50 times by the users who has rated the books more than 250 times.

In [45]:
# Lets pivot the table:

pt = final_ratings.pivot_table(index = 'Book-Title', columns = 'User-ID', values = 'Book-Rating')

pt

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN    10.0     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     0.0     NaN     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     0.0     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   NaN  ...    10.0   
1st to Die: A Novel                                    9.0  ...     NaN   
2nd Chance                                             0.0  ...     NaN   
4 Blondes                                              NaN  ...     NaN   
A Bend in the Road                                     NaN  ...     NaN   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     NaN   
You Belong To Me                                       NaN  ...     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN  ...     NaN   
Zoya                                                   NaN  ...     NaN   
\O\" Is for Outlaw"                                    

Lets fill missing values with 0.

In [46]:
pt.fillna(0, inplace=True)

In [47]:
pt

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   0.0  ...    10.0   
1st to Die: A Novel                                    9.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
A Bend in the Road                                     0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

In [48]:
from sklearn.metrics.pairwise import cosine_similarity

In [49]:
similarity_score = cosine_similarity(pt)

In [50]:
similarity_score.shape

(706, 706)

In [51]:
similarity_score

array([[1.        , 0.10255025, 0.01220856, ..., 0.12110367, 0.07347567,
        0.04316046],
       [0.10255025, 1.        , 0.2364573 , ..., 0.07446129, 0.16773875,
        0.14263397],
       [0.01220856, 0.2364573 , 1.        , ..., 0.04558758, 0.04938579,
        0.10796119],
       ...,
       [0.12110367, 0.07446129, 0.04558758, ..., 1.        , 0.07085128,
        0.0196177 ],
       [0.07347567, 0.16773875, 0.04938579, ..., 0.07085128, 1.        ,
        0.10602962],
       [0.04316046, 0.14263397, 0.10796119, ..., 0.0196177 , 0.10602962,
        1.        ]])

In [59]:
'''
Lets build a function which will taek book_name as the parameters 
and return 5 books that has similarity scores with that particluar based on the above array.

'''
def recommend(book_name):
    #fetching_index
    index = np.where(pt.index == book_name)[0][0]
    # Getching 5 books which has highest similarity scores.
    similar_items = sorted(list(enumerate(similarity_score[index])), key = lambda x : x[1], reverse = True)[1:6]
    
    for i in similar_items:
        print(pt.index[i[0]])   

In [62]:
recommend('1984')

Animal Farm
The Handmaid's Tale
Brave New World
The Vampire Lestat (Vampire Chronicles, Book II)
The Hours : A Novel


### Recommen function recommends the 5 films that are very similiar to the movie which is being passed to the function, this is a simple way of building a recommender systems.

In [65]:
import pickle
pickle.dump(popular_df,open('popular.pkl','wb'))

In [66]:
popular_df.head()

Book-Title    Book-Author  \
0   Harry Potter and the Prisoner of Azkaban (Book 3)  J. K. Rowling   
3        Harry Potter and the Goblet of Fire (Book 4)  J. K. Rowling   
5      Harry Potter and the Sorcerer's Stone (Book 1)  J. K. Rowling   
9   Harry Potter and the Order of the Phoenix (Boo...  J. K. Rowling   
13   Harry Potter and the Chamber of Secrets (Book 2)  J. K. Rowling   

                                          Image-URL-M  num_ratings  \
0   http://images.amazon.com/images/P/0439136350.0...          428   
3   http://images.amazon.com/images/P/0439139597.0...          387   
5   http://images.amazon.com/images/P/0590353403.0...          278   
9   http://images.amazon.com/images/P/043935806X.0...          347   
13  http://images.amazon.com/images/P/0439064872.0...          556   

    avg_ratings  
0      5.852804  
3      5.824289  
5      5.737410  
9      5.501441  
13     5.183453